ggu/main.c

```diff
@@ -16,11 +16,401 @@
+		process_single_policy(&policy, ggu_conf);
+	}
+
+	/* Process the IPv4 granted decisions. */
+	for (j = 0; j < gguhdr->n3; j++) {
+		memset(&policy, 0, sizeof(policy));
+		policy.state = GK_GRANTED;
+		policy.flow.proto = ETHER_TYPE_IPv4;
+		rte_memcpy(&policy.flow.f.v4, policy_ptr,
+			sizeof(policy.flow.f.v4));
+		policy_ptr += sizeof(policy.flow.f.v4);
+		rte_memcpy(&policy.params.u.granted, policy_ptr,
+			sizeof(policy.params.u.granted));
+		policy_ptr += sizeof(policy.params.u.granted);
+		process_single_policy(&policy, ggu_conf);
+	}
+
+	/* Process the IPv6 granted decisions. */
+	for (j = 0; j < gguhdr->n4; j++) {
+		memset(&policy, 0, sizeof(policy));
+		policy.state = GK_GRANTED;
+		policy.flow.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&policy.flow.f.v6, policy_ptr,
+			sizeof(policy.flow.f.v6));
+		policy_ptr += sizeof(policy.flow.f.v6);
+		rte_memcpy(&policy.params.u.granted, policy_ptr,
+			sizeof(policy.params.u.granted));
+		policy_ptr += sizeof(policy.params.u.granted);
+		process_single_policy(&policy, ggu_conf);
+	}
+
+free_packet:
+	rte_pktmbuf_free(pkt);
+}
+
+static int
+ggu_proc(void *arg)
+{
+	int ret;
+	uint32_t lcore = rte_lcore_id();
+	struct ggu_config *ggu_conf = (struct ggu_config *)arg;
+	uint8_t port_in = ggu_conf->net->back.id;
+	uint16_t rx_queue = ggu_conf->rx_queue_back;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit is running at lcore = %u\n", lcore);
+
+	/* Wait for network devices to start. */
+	while (ggu_conf->net->configuring)
+		;
+
+	/*
+	 * Setup the ntuple filter that assign the GK-GT packets
+	 * to its queue for both IPv4 and IPv6 addresses.
+	 * Accordnig to the DPDK documentation, any functions from
+	 * the rte_ethdev.h library should be called after the 
+	 * network devices are started. So as the ntuple filters functions.
+	 */
+	ret = ntuple_filter_add(port_in,
+		ggu_conf->net->back.ip4_addr.s_addr,
+		ggu_conf->ggu_src_port,
+		ggu_conf->ggu_dst_port,
+		ggu_conf->rx_queue_back);
+	if (ret < 0) {
+		exiting = true;
+		goto out;
+	}
+
+	while (likely(!exiting)) {
+		uint16_t i;
+		uint16_t num_rx;
+		struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+
+		/* Load a set of GK-GT packets from the back NIC. */
+		num_rx = rte_eth_rx_burst(port_in, rx_queue, bufs,
+			GATEKEEPER_MAX_PKT_BURST);
+		if (unlikely(num_rx == 0))
+			continue;
+
+		for (i = 0; i < num_rx; i++)
+			process_single_packet(bufs[i], ggu_conf);
+	}
+
+out:
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"ggu: the GK-GT unit at lcore = %u is exiting\n", lcore);
+	return cleanup_ggu(ggu_conf);
+}
 
 int
-run_ggu(__attribute__((unused)) const struct ggu_config *ggu_conf)
+run_ggu(struct net_config *net_conf,
+	struct gk_config *gk_conf, struct ggu_config *ggu_conf)
 {
-	/* TODO Initialize and run GK-GT Unit functional block. */
+	int ret;
+
+	if (ggu_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (net_conf == NULL || gk_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	ggu_conf->net = net_conf;
+	ggu_conf->gk = gk_conf;
```

Comment:
We have to increment the reference count to `gk_conf`, that is, calling `rte_atomic32_init(&gk_conf->ref_cnt);`. But instead of calling directly, the GK block should define `gk_conf_hold()` and `gk_conf_put()` to enable other blocks to do so.

Special care must be taken below in this function in case of an error does occur since the reference must be cleared. Nulling `ggu_conf->net` and `ggu_conf->gk` is also recommended.

Remember that `cleanup_ggu()` will need to call `gk_conf_put()` as well.